lib/adapters/show-document-adapter.ts

```diff
@@ -0,0 +1,59 @@
+import { shell } from "electron"
+import { LanguageClientConnection, ShowDocumentParams, ShowDocumentResult } from "../languageclient"
+import { TextEditor } from "atom"
+import Convert from "../convert"
+
+/** Public: Adapts the window/showDocument command to Atom's text editors or external programs. */
+export default class ShowDocumentAdapter {
+  /** {@inheritDoc attach} */
+  public attach(arg: Parameters<typeof attach>[0]): ReturnType<typeof attach> {
+    attach(arg)
+  }
+  /** {@inheritDoc showDocument} */
+  public onShowDocument(...args: Parameters<typeof showDocument>): ReturnType<typeof showDocument> {
+    return showDocument(...args)
+  }
+}
+
+/**
+ * Public: Attach to a {LanguageClientConnection} to recieve requests to show documents.
+ */
+export function attach(connection: LanguageClientConnection): void {
+  connection.onShowDocument(showDocument)
+}
+
+/**
+ * Public: show documents inside Atom text editor or in external programs
+ *
+ * @param params The {ShowDocumentParams} received from the language server
+ *   indicating the document to be displayed as well as other metadata.
+ * @returns {Promise<ShowDocumentResult>} with a `success: boolean` property specifying if the operation was sucessful
+ * {@inheritDoc ShowDocumentParams}
+ */
+export async function showDocument(params: ShowDocumentParams): Promise<ShowDocumentResult> {
+  try {
+    if (!params.external) {
+      // open using atom.workspace
+      const view = await atom.workspace.open(Convert.uriToPath(params.uri), {
+        activateItem: params.takeFocus,
+        activatePane: params.takeFocus,
+        pending: true,
+        initialLine: params.selection?.start.line ?? 0,
+        initialColumn: params.selection?.start.character ?? 0,
+      })
+      if (view === undefined) {
```

Comment:
I'm not convinced that this won't return null in the future. It should always be truthy if it exists.

```suggestion
      if (!view) {
```